yarn-project/txe/src/oracle/txe_oracle.ts

```diff
@@ -544,33 +600,260 @@ export class TXE implements TypedOracle {
-    _isDelegateCall: boolean,
+  public async getDebugFunctionName(address: AztecAddress, selector: FunctionSelector): Promise<string | undefined> {
+    const instance = await this.contractDataOracle.getContractInstance(address);
+    if (!instance) {
+      return undefined;
+    }
+    const artifact = await this.contractDataOracle.getContractArtifact(instance!.contractClassId);
+    if (!artifact) {
+      return undefined;
+    }
+
+    const f = artifact.functions.find(f =>
+      FunctionSelector.fromNameAndParameters(f.name, f.parameters).equals(selector),
+    );
+    if (!f) {
+      return undefined;
+    }
+
+    return `${artifact.name}:${f.name}`;
+  }
+
+  async executePublicFunction(
+    targetContractAddress: AztecAddress,
+    functionSelector: FunctionSelector,
+    args: Fr[],
+    callContext: CallContext,
+  ) {
+    const fnName =
+      (await this.getDebugFunctionName(targetContractAddress, functionSelector)) ??
+      `${targetContractAddress}:${functionSelector}`;
+
+    this.logger.debug(`[AVM] Executing public external function ${fnName}.`);
+    const timer = new Timer();
+    const startGas = Gas.test();
+    const hostStorage = new HostStorage(
+      new TXEPublicStateDB(this),
+      new ContractsDataSourcePublicDB(new TXEPublicContractDataSource(this)),
+      new WorldStateDB(this.trees.asLatest()),
+    );
+
+    const worldStateJournal = new AvmPersistableStateManager(hostStorage);
+    for (const nullifier of this.noteCache.getNullifiers(targetContractAddress)) {
+      worldStateJournal.nullifiers.cache.appendSiloed(new Fr(nullifier));
+    }
+    worldStateJournal.trace.accessCounter = callContext.sideEffectCounter;
+    const execution = {
+      contractAddress: targetContractAddress,
+      functionSelector,
+      args,
+      callContext,
+    };
+    const header = Header.empty();
+    header.state = await this.trees.getStateReference(true);
+    header.globalVariables.blockNumber = new Fr(await this.getBlockNumber());
+    header.state.partial.nullifierTree.root = Fr.fromBuffer(
+      (await this.trees.getTreeInfo(MerkleTreeId.NULLIFIER_TREE, true)).root,
+    );
+    header.state.partial.noteHashTree.root = Fr.fromBuffer(
+      (await this.trees.getTreeInfo(MerkleTreeId.NOTE_HASH_TREE, true)).root,
+    );
+    header.state.partial.publicDataTree.root = Fr.fromBuffer(
+      (await this.trees.getTreeInfo(MerkleTreeId.PUBLIC_DATA_TREE, true)).root,
+    );
+    header.state.l1ToL2MessageTree.root = Fr.fromBuffer(
+      (await this.trees.getTreeInfo(MerkleTreeId.L1_TO_L2_MESSAGE_TREE, true)).root,
+    );
+    const executionEnv = createAvmExecutionEnvironment(
+      execution,
+      header,
+      GlobalVariables.empty(),
+      GasSettings.default(),
+      Fr.ZERO,
+    );
+
+    const machineState = new AvmMachineState(startGas);
+    const avmContext = new AvmContext(worldStateJournal, executionEnv, machineState);
+    const simulator = new AvmSimulator(avmContext);
```

Comment:
We should talk about this. CCing @dbanks12 as well.

You are using VERY internal things like the transitional adaptors, and accessing things like `avmContext.persistableState.publicStorage` which are only transitionally public.

This is really not supported and I think we cannot commit to it. The interfaces are still changing, especially things that you now export from `public`. 

Much of the logic as well is what is now in the public executor which would be duplicated. Usually we have freedom to modify what we want there because it's siloed.

I'm very much against this in it's current form and I'm not sure we can provide a better form in the short term.